### ISSUE TITLE
OpenCL code needs to be cleaned up

### ISSUE THREAD

**Issue by mathuin** [mathuin/TopoMC#20]
First step is to document what the code currently does.

Second step is to write some sort of test harness, to confirm that the code does what the documentation says it does.

Any performance enhancements come after this bit.

**Comment by mathuin** [comment]
I threw away the old OpenCL code.

I wrote a replacement for the cKDTree/Invdisttree stuff in OpenCL.  It runs from ten to sixty times faster than the old code and handles larger worlds than the old code.  It has a test harness.  The error rate (difference in results between old code and new code) is between 0.05% and 0.5% on average.  This is good.

I now think that OpenCL could improve the speed of building the three-dimensional arrays from which worlds are generated.  Before I can test that theory, I need to abstract out the OpenCL portion of the CLIDT class into its own class.  Once that is done and I confirm that everything still works, I'll release it into the develop branch for others to test.

**Comment by mathuin** [comment]
Okay, nearly every part of ./PrepRegion.py has been improved with OpenCL.  I'm running a bunch of showcase sites through the system now to see how performance improves, but I'm seeing some improvement even when not using OpenCL as a result of refactoring the code.  Exciting!

Most of the changes have error rates below 1% (less than one out of every hundred values differs from the 'old way' of doing things).  The bathy part has an error rate of 4.6% on Block Island but it doesn't "look wrong" so I'm letting it slide -- especially since it does in 0.1 second what took 2 seconds before!  The last thing I want to change in ./PrepRegion.py is replacing the gdalwarp call for elevation with OpenCL and cKDTree/Invdisttree.  If it's faster for both, I'll leave it that way and commit the feature to develop.  If it's only faster for OpenCL, then I'll rework things first.

The only thing I don't really like about the new feature is that it looks like a horrible violation of Don't Repeat Yourself.  Unfortunately the differences between bits seem both subtle and necessary so I'm stuck with it.

Other than that, it's a massive improvement.  Yay.

**Comment by mathuin** [comment]
The gdalwarp call stays for now, until the Python API gets warping.

I ran the new code against fifteen different regions, and came up with an average speed-up of over 50% with the new OpenCL, and over 35% without!  Wow!  I'm merging this into develop and will probably roll out a release once I add OpenCL support to ./BuildRegion.py.

**Comment by mathuin** [comment]
I have added the code to the develop branch.  On further inspection, there's a problem with the non-OpenCL version of the elev code -- it places some interesting artifacts in some obvious places.  I need to examine this in more detail.

**Comment by mathuin** [comment]
This next commit fixes most of the artifacts.  I'm okay with the rest.

Next stop: ./BuildRegion.py!

**Comment by mathuin** [comment]
I'm going to use a separate issue for adding OpenCL to BuildRegion.py -- since the original goals of this issue were achieved, I'm going to close it.
